File: sdks/typescript/src/core/emergency-revocation.ts
/**
 * DCP v2.0 Emergency Revocation (Gap #13).
 *
 * Pre-registered revocation token allows revoking all agent keys without
 * requiring a private key signature. The human stores the secret offline;
 * revealing the pre-image proves authorization.
 *
 * Token lifecycle:
 *   1. At identity creation: secret = random(32), token = sha256(secret)
 *   2. Token is stored in AgentPassportV2.emergency_revocation_token
 *   3. Secret is stored offline by the human
 *   4. On emergency: human reveals secret, gateway verifies sha256(secret) == token
 */

import { randomBytes, createHash } from 'crypto';
import type { EmergencyRevocation } from '../types/v2.js';

export interface EmergencyRevocationTokenPair {
  /** The secret to store offline (64 hex chars). */
  revocation_secret: string;
  /** The commitment to embed in the passport: "sha256:<hex>". */
  emergency_revocation_token: string;
}

/**
 * Generate a revocation secret and its commitment token.
 * The secret MUST be stored offline; the token goes in the passport.
 */
export function generateEmergencyRevocationToken(): EmergencyRevocationTokenPair {
  const secretBytes = randomBytes(32);
  const secret = secretBytes.toString('hex');
  const hash = createHash('sha256').update(secretBytes).digest('hex');
  return {
    revocation_secret: secret,
    emergency_revocation_token: `sha256:${hash}`,
  };
}

/**
 * Verify that a revocation secret matches the commitment token in a passport.
 *
 * @param revocationSecret - The 64-char hex pre-image
 * @param commitmentToken - The "sha256:<hex>" token from the passport
 * @returns true if sha256(secret) matches the commitment
 */
export function verifyEmergencyRevocationSecret(
  revocationSecret: string,
  commitmentToken: string,
): boolean {
  if (!commitmentToken.startsWith('sha256:')) return false;
  const expectedHex = commitmentToken.slice('sha256:'.length);
  const secretBytes = Buffer.from(revocationSecret, 'hex');
  if (secretBytes.length !== 32) return false;
  const actualHex = createHash('sha256').update(secretBytes).digest('hex');
  return actualHex === expectedHex;
}

/**
 * Build an EmergencyRevocation request object.
 */
export function buildEmergencyRevocation(params: {
  agentId: string;
  humanId: string;
  revocationSecret: string;
}): EmergencyRevocation {
  return {
    type: 'emergency_revocation',
    agent_id: params.agentId,
    human_id: params.humanId,
    revocation_secret: params.revocationSecret,
    timestamp: new Date().toISOString(),
    reason: 'key_compromise_emergency',
  };
}

File: sdks/python/dcp_ai/v2/emergency_revocation.py
"""
DCP v2.0 Emergency Revocation — Python port.

Pre-registered revocation token allows revoking all agent keys without
requiring a private-key signature. The human keeps `revocation_secret`
offline; revealing its pre-image proves authorisation.
"""

from __future__ import annotations

import hashlib
import hmac
import secrets
from typing import Any

from dcp_ai.v2.session_nonce import generate_session_expiry  # just to register shared imports


def generate_emergency_revocation_token() -> dict[str, str]:
    """Generate a (revocation_secret, emergency_revocation_token) pair.

    The secret MUST be stored offline; the token is embedded in the passport.
    Returns {'revocation_secret': hex64, 'emergency_revocation_token': 'sha256:<hex64>'}.
    """
    secret_bytes = secrets.token_bytes(32)
    secret_hex = secret_bytes.hex()
    commit = hashlib.sha256(secret_bytes).hexdigest()
    return {
        "revocation_secret": secret_hex,
        "emergency_revocation_token": f"sha256:{commit}",
    }


def verify_emergency_revocation_secret(
    revocation_secret: str,
    commitment_token: str,
) -> bool:
    """Return True iff sha256(secret) matches the `sha256:<hex>` commitment token."""
    if not commitment_token.startswith("sha256:"):
        return False
    try:
        secret_bytes = bytes.fromhex(revocation_secret)
    except ValueError:
        return False
    if len(secret_bytes) != 32:
        return False
    expected_hex = commitment_token[len("sha256:"):]
    actual = hashlib.sha256(secret_bytes).hexdigest()
    return hmac.compare_digest(actual, expected_hex)


def build_emergency_revocation(
    *,
    agent_id: str,
    human_id: str,
    revocation_secret: str,
) -> dict[str, Any]:
    """Build an EmergencyRevocation request object ready to send to the gateway."""
    from datetime import datetime, timezone
    ts = datetime.now(timezone.utc)
    ts_iso = ts.strftime("%Y-%m-%dT%H:%M:%S.") + f"{ts.microsecond // 1000:03d}Z"
    return {
        "type": "emergency_revocation",
        "agent_id": agent_id,
        "human_id": human_id,
        "revocation_secret": revocation_secret,
        "timestamp": ts_iso,
        "reason": "key_compromise_emergency",
    }


# Expose so static analysers don't flag it as unused; we imported it to make
# absolutely sure the shared expiry helper stays in the public v2 surface.
_ = generate_session_expiry
